src/nagini_translation/translators/expression.py

```diff
@@ -717,6 +717,46 @@ def translate_static_field_access(self, field: PythonGlobalVar,
                 ctx.current_function.call_deps.add((node, receiver, ctx.module))
         return res
 
+    def translate_adt_decons(self, decons: PythonClass, node: ast.Attribute,
+                             pos: Position, ctx: Context) -> StmtsAndExpr:
+        """
+        Deconstructs the ADT into the selected component, boxing the result
+        if the projected component's type is ADT (recursive composition).
+        """
+        info = self.no_info(ctx)
+        adt_name = decons.adt_def.name + '_ADT'
+        adt_prefix = decons.adt_def.name + '_'
+        adt_type = self.viper.DomainType(adt_name, {}, [])
+
+        # If ADT is casted, ignore the casting since 'decons' already reflects
+        # the casted type
+        if isinstance(node.value , ast.Call) and node.value.func.id == 'cast':
+            stmt, adt_obj = self.translate_expr(node.value.args[1], ctx)
+        else:
+            stmt, adt_obj = self.translate_expr(node.value, ctx)
+
+        # Unbox ADT to be deconstructed
+        unbox_func = self.viper.FuncApp('unbox_' + adt_name, [adt_obj], pos,
+                                        info, adt_type)
+
+        # Calculate return type
+        if decons.fields[node.attr].type == decons.adt_def:
```

Comment:
Same question as above, why do we need special treatment for this? Would things also work if we had an ADT with a field whose type is a different ADT?

Reply:
The idea is that all "foreign" types stored in the ADT are Ref (e.g. int, str, primitive types or even custom types), whereas the ADT stored in an ADT has its own native type (recursively). A different ADT is considered a "foreign" type, therefore will be stored as a Ref. When deconstructing (accessing an attribute), the proper return type should be considered: either a Ref or the ADT own type, depending on the attribute being considered. When appropriate, the returned Ref will be unboxed to its corresponding original type, symmetrically to when the data was stored in the ADT (boxed).

Reply:
Okay, sounds good